crates/edr_generic/src/spec.rs

```diff
@@ -37,6 +41,66 @@ impl ChainSpec for GenericChainSpec {
     type Context = ();
     type HaltReason = l1::HaltReason;
     type SignedTransaction = crate::transaction::SignedWithFallbackToPostEip155;
+    type BlockConstructor = GenericBlockConstructor;
+}
+
+/// Generic chain definition for `BlockEnvConstructor`
+pub struct GenericBlockConstructor;
+
+impl BlockEnvConstructor<Header, BlockEnv> for GenericBlockConstructor {
+    fn new_block_env(header: &Header, hardfork: l1::SpecId) -> BlockEnv {
+        let is_prague = hardfork >= l1::SpecId::PRAGUE;
+        BlockEnv {
+            number: header.number,
+            beneficiary: header.beneficiary,
+            timestamp: header.timestamp,
+            difficulty: header.difficulty,
+            basefee: header.base_fee_per_gas.map_or(0u64, |base_fee| {
+                base_fee.try_into().expect("base fee is too large")
+            }),
+            gas_limit: header.gas_limit,
+            prevrandao: if hardfork >= l1::SpecId::MERGE {
+                Some(header.mix_hash)
+            } else {
+                None
+            },
+            blob_excess_gas_and_price: header
+                .blob_gas
+                .as_ref()
+                .map(|BlobGas { excess_gas, .. }| {
+                    eip4844::BlobExcessGasAndPrice::new(*excess_gas, is_prague)
+                })
+                .or_else(|| Some(eip4844::BlobExcessGasAndPrice::new(0u64, is_prague))),
+        }
+    }
+}
+
+impl BlockEnvConstructor<PartialHeader, BlockEnv> for GenericBlockConstructor {
+    fn new_block_env(header: &PartialHeader, hardfork: l1::SpecId) -> BlockEnv {
+        let is_prague = hardfork >= l1::SpecId::PRAGUE;
+        BlockEnv {
+            number: header.number,
+            beneficiary: header.beneficiary,
+            timestamp: header.timestamp,
+            difficulty: header.difficulty,
+            basefee: header.base_fee.map_or(0u64, |base_fee| {
+                base_fee.try_into().expect("base fee is too large")
+            }),
+            gas_limit: header.gas_limit,
+            prevrandao: if hardfork >= l1::SpecId::MERGE {
+                Some(header.mix_hash)
+            } else {
+                None
+            },
+            blob_excess_gas_and_price: header
+                .blob_gas
+                .as_ref()
+                .map(|BlobGas { excess_gas, .. }| {
+                    eip4844::BlobExcessGasAndPrice::new(*excess_gas, is_prague)
+                })
+                .or_else(|| Some(eip4844::BlobExcessGasAndPrice::new(0u64, is_prague))),
```

Comment:
We should still check whether the `BlobExcessGasAndPrice` is required for this particular hardfork.

Could you also add a test that validates that we don't default the `BlobExcessGasAndPrice` for pre-Cancun hardforks? Similarly for L1